sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py

```diff
@@ -0,0 +1,150 @@
+
+from io import BytesIO
+
+from avro.datafile import DataFileReader
+from avro.io import DatumReader
+
+from ._models import QuickQueryError
+
+
+class QuickQueryReader(object):  # pylint: disable=too-many-instance-attributes
+    """A streaming object to read quick query result.
+
+    :ivar str name:
+        The name of the blob for the quick query request.
+    :ivar str container:
+        The name of the container where the blob is.
+    :ivar dict response_headers:
+        The response_headers of the quick query request.
+    :ivar int total_bytes:
+        The size of the total data in the stream.
+    """
+
+    def __init__(
+        self,
+        client=None,
+        name=None,
+        container=None,
+        progress_callback=None,
+        encoding=None,
+        **kwargs
+    ):
+        self.name = name
+        self.container = container
+        self.response_headers = None
+        self.total_bytes = None
+        self.bytes_processed = 0
+        self._progress_callback = progress_callback
+        self._client = client
+        self._request_options = kwargs
+        self._encoding = encoding
+
+    def __len__(self):
+        return self.size
+
+    def readall(self):
+        """Return all quick query results.
+
+        This operation is blocking until all data is downloaded.
+        :rtype: bytes
+        """
+        stream = BytesIO()
+        self.readinto(stream)
+        data = stream.getvalue()
+        if self._encoding:
+            return data.decode(self._encoding)
+        return data
+
+    def readinto(self, stream):
+        """Download the quick query result to a stream.
+
+        :param stream:
+            The stream to download to. This can be an open file-handle,
+            or any writable stream.
+        :returns: None
+        """
+        headers, raw_response_body = self._client.blob.quick_query(**self._request_options)
+        self.response_headers = headers
+        self._parse_quick_query_result(raw_response_body, stream, progress_callback=self._progress_callback)
+
+    def _parse_quick_query_result(self, raw_response_body, stream, progress_callback=None):
+        parsed_results = DataFileReader(QuickQueryStreamer(raw_response_body), DatumReader())
+        for parsed_result in parsed_results:
+            if parsed_result.get('data'):
+                stream.write(parsed_result.get('data'))
+                continue
+
+            self.bytes_processed = parsed_result.get('bytesScanned')
+            self.total_bytes = parsed_result.get('totalBytes')
+            fatal = QuickQueryError(parsed_result['name'],
+                                    parsed_result['fatal'],
+                                    parsed_result['description'],
+                                    parsed_result['position']) if parsed_result.get('fatal') else None
+            if progress_callback:
+                progress_callback(fatal, self.bytes_processed, self.total_bytes)
```

Comment:
We should remove this progress callback - instead we can make this object iterable, so that if a user needs this information that can iterate over the data in the result.

Reply:
fatal, bytes_processed and total bytes are service defined schema, they are mixed with data in query response body. That means when we use avro to parse the query response body, avro could give us some data(not guaranteed a full record), fatal, bytes_processed or total bytes. If we parse the record for users, we bytes_processed will not make sense, because we can only give user the full record even more bytes for the next record has been returned by arvo parser. 

Besides the record could be JSON, or CSV type, the record separator and column separator etc. are based on what user specified which make it more difficult to extract the result from 

Also currently query doesn't allow us to do range download.. Once we trigger query, it will keep returning data until there's no more result. So we cannot use ItemPaged....

We had a discussion across all platforms and think make it has a similar interface as download, we will have further discussion if you feel strongly about making it iterable.

Reply:
Thoughts @annatisch ?